protocol/test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol

```diff
@@ -0,0 +1,165 @@
+      daiInterestRateModel,
+      borrowDaiAmount,
+      BORROW_CEILING
+    ); // 7.77 %
+
+    expectedDaiDebt = approxInterest(
+      borrowDaiAmount,
+      expectedDaiInterestRate,
+      365 days
+    ); // ~54k
+
+    expectedUtilizationRation = utilizationRatio(
+      expectedDaiDebt,
+      BORROW_CEILING
+    ); // ~54%
+  }
+
+  function utilizationRatio(
+    uint256 borrowed,
+    uint256 cap
+  ) internal pure returns (uint256) {
+    return (borrowed * 1e18) / cap;
+  }
+
+  function calculateInterestRate(
+    IInterestRateModel model,
+    uint256 borrowed,
+    uint256 cap
+  ) internal view returns (uint96) {
+    return model.calculateInterestRate(utilizationRatio(borrowed, cap));
+  }
+
+  function approxInterest(
+    uint256 principal,
+    uint96 rate,
+    uint256 age
+  ) internal pure returns (uint256) {
+    // Approxmiate as P * (1 + r/365 days)^(age)
+    // ie compounding every 1 second (almost but not quite continuous)
+    uint256 onePlusRate = uint256(rate / 365 days + 1e18);
+
+    return ud(principal).mul(ud(onePlusRate).powu(age)).unwrap();
+  }
+}
+
+contract LinearWithKinkInterestRateModelTestModifiers is
+  LinearWithKinkInterestRateModelTestBase
+{
+  event InterestRateParamsSet(
+    uint256 _baseInterestRate,
+    uint256 _maxInterestRate,
+    uint256 _kinkUtilizationRatio,
+    uint256 _kinkInterestRate
+  );
+
+  using SafeCast for uint256;
+
+  function test_accessSetRateParams() public {
+    // no access for alice
+    {
+      vm.startPrank(alice);
+      vm.expectRevert(
+        abi.encodeWithSelector(CommonEventsAndErrors.InvalidAccess.selector)
+      );
+      mock.setRateParams(2e18 / 100, 75e18 / 100, 50e18 / 100, 10e18 / 100);
+      vm.stopPrank();
+    }
+
+    // access for executor
+    {
+      vm.startPrank(executor);
+      vm.expectEmit();
+      emit InterestRateParamsSet(2e18 / 100, 75e18 / 100, 50e18 / 100, 10e18 / 100);
+      mock.setRateParams(2e18 / 100, 75e18 / 100, 50e18 / 100, 10e18 / 100);
+    }
+  }
+
+  function test_getRateParams() public {    
+    uint80 testBaseInterestRate = uint256(2e18 / 100).encodeUInt80();
+    uint80 testMaxInterestRate = uint256(80e18 / 100).encodeUInt80();
+    uint80 testKinkInterestRate = uint256(81e18 / 100).encodeUInt80();
+    uint256 testKinkUtilizationRatio = uint256(50e18 / 100);
+    
+    // get initial params
+    (
+      uint256 baseInterestRate,
+      uint256 maxInterestRate,
+      uint256 kinkInterestRate,
+      uint256 kinkUtilizationRatio
+    ) = mock.rateParams();
+    
+    assertEq(baseInterestRate, testBaseInterestRate);
+    assertEq(maxInterestRate, testMaxInterestRate);
+    assertEq(kinkInterestRate, testKinkInterestRate);
+    assertEq(kinkUtilizationRatio, testKinkUtilizationRatio);
+  }
+
+  function test_computeInterestRateImpl() public {
+    uint256 utilizationRation = mock.computeInterestRateImpl(expectedUtilizationRation);
+
+    assertEq(utilizationRation, expectedDaiInterestRate);
+
+  }
```

Comment:
@frontier159 this is throwing error here, I think I don't need to test `computeInterestRateImpl`? I just realized it is an internal function 😅

```
Compiler run failed:
Error (9582): Member "computeInterestRateImpl" not found or not visible after argument-dependent lookup in contract Mock.
   --> test/forge/v2/interestRate/LinearWithKinkInterestRateModel.t.sol:160:33:
    |
160 |     uint256 utilizationRation = mock.computeInterestRateImpl(expectedUtilizationRation);
```

Reply:
The visibility is just like in OO classes. Private == only visible to that contract, internal == visible to that contract and any contact which inherits it